src/internal/use-smart-update.ts

```diff
@@ -0,0 +1,40 @@
+import { useCallback, useLayoutEffect, useRef, useState } from 'react';
+import { asapScheduler } from 'rxjs';
+
+function useForceUpdate(): () => void {
+	const [, dispatch] = useState<{}>(Object.create(null));
+
+	const forceUpdate = useCallback((): void => {
+		dispatch(Object.create(null));
+	}, [dispatch]);
+	return forceUpdate;
+}
+
+function useSmartUpdate() {
+	const scheduledUpdateRef = useRef<undefined | (() => void)>();
+	const forceUpdate = useForceUpdate();
+
+	useLayoutEffect(() => {
+		// If a rerender has occurred for some other reason or the component
+		// has unmounted then deschedule any pending forced updates.
+		return () => {
+			scheduledUpdateRef.current = undefined;
+		};
+	});
```

Comment:
This check should probably belong in `useWatchObservable`. Instead of doing this in useSmartUpdate we could have the returned function return another function for desceduling  pending updates